### ISSUE TITLE
Add support for commitCharacters property

### ISSUE THREAD

**Issue by Shatur95** [nvim-lua/completion-nvim#128]
**Is your feature request related to a problem? Please describe.**
Currently all snippets auto-expanded. This is useful. But sometimes I do not need expansion, for example:

```cpp
object.my_function(argument1, argument2);
```

If I want to change `my_function` I can't use completion because it will be expanded automatically and I will have something like the following:

```cpp
object.another_function(snippet_argument1, snippet_argument2)(argument1, argument2);
```

**Describe the solution you'd like**
So, I suggest to make this behavior more smart. For example, do not expand function if next symbol is not a space. This is how it works in most IDE's. The LSP spec has the `CompletionItem#commitCharacters` that is the same concept. I have [a discussion](https://github.com/hrsh7th/vim-vsnip-integ/issues/22) with @hrsh7th an he says that such thing need to be implemented in completion or LSP engine.

**Describe alternatives you've considered**
[There was an attempt](https://github.com/hrsh7th/vim-vsnip-integ/pull/23) for a manual check for a next character, but it causes conflicts with autopair-like plugins :(

**Comment by haorenW1025** [comment]
Hmm I've looked into it and it might be harder to implement this than I initially thought, cause If the snippets are not expand it's just a bunch of junk words ( like `object.my_function(v:argument1, v:argument2)`. So if I want to support this I should trim words in the completion stage.

**Comment by Shatur95** [comment]
Maybe snippet plugin should provide an API to access to unexpanded words?

**Comment by haorenW1025** [comment]
I'll try doing that in #131. Hopefully we can get a good solution:)

**Comment by Shatur95** [comment]
You meant #132, not #131, right?
Looks very cool! Let me test it :)

**Comment by haorenW1025** [comment]
Ah yeah sorry for the mistype😥😥

**Comment by haorenW1025** [comment]
@Shatur95 Do you know which lsp have commitCharacter support? I try to check with several lsp I know but seems most of them doesn't support that.

**Comment by Shatur95** [comment]
It looks like this feature was added to [clangd](https://github.com/llvm/llvm-project/commit/2a095ff6f5028b7612dae50358a2f6939b6cdc9f) recently. But If I understand correctly, this will be available only in nightly version :(

**Comment by hrsh7th** [comment]
Hm... I searched the server that supports `commitCharacters` but found the few servers only.

I can provide `disable_automatic_expansion` flag to vsnip.

The completion-nvim handle CompleteDone by itself so when this flag enabled, the snippet will be expanded only when the user press `<CR>`.

**Comment by hrsh7th** [comment]
But completion-nvim has already removed vsnip-integ's autocmd, so it should be....

**Comment by haorenW1025** [comment]
@hrsh7th That's already the expected behavior in completion-nvim because I think if I handle completeDone by myself i don't need the autocommands.

@Shatur95 That's fine, I'll look into the nightly version.

**Comment by haorenW1025** [comment]
@hrsh7th I do have something I wanna ask you, is there a way to get the word without snippets placeholders in LSP? For example I want to get `object.my_function` instead of `object.my_function(v:argument1, v:argument2)`. Right now I manually trims the snippets placeholder but it doesn't seem very smart.

**Comment by hrsh7th** [comment]
In my knowledge, Many LSP client does it via trimming insertText.
So I think current approach is not bad.

But my own solution is different a bit (Actually, I have [my own LSP client](https://github.com/hrsh7th/vim-lamp) 😅.

My solution is using `CompletionItem.label` as `v:completed_item.word` when CompletionItem's insertTextFormat is `snippet`.

Sorry, I don't know which solution is reasonable in LSP spec.

**Comment by hrsh7th** [comment]
My solution is [here](https://github.com/hrsh7th/vim-lamp/blob/master/autoload/lamp/feature/completion.vim#L79).

coc.nvim's solution is maybe [here](https://github.com/neoclide/coc.nvim/blob/5d219ebd25569c297c687a7625a0227363528f3c/src/util/complete.ts#L77).

**Comment by haorenW1025** [comment]
@hrsh7th That's very helpful! Thanks a lot for the information😉

**Comment by hrsh7th** [comment]
I think this problem exists only in vim because VSCode has no inserting `v:completed_item.word` when the user chooses a candidate.

So we have to consider how to solve it by ourselves maybe...

**Comment by haorenW1025** [comment]
Another question here, so the commitCharacters is the characters that *will* allow snippets expansion, am I correct? Cause when I look into the `allcommitCharacters` in clangd it's defined as `allCommitCharacters = " \t()[]{}<>:;,+-/*%^&#?.=\"'|"`. I don't think it help solving @Shatur95 issue if it is....

**Comment by Shatur95** [comment]
Strange...
Maybe the `commitCharacters` is the characters that use smart expansion?..
In QtCreator `window->a|()` (where `|` is cursor) becomes `window->addWidget(|)` after completion.

**Comment by haorenW1025** [comment]
Hmm then that make the `" "` and `"\t"` a little weird...Also I look at the spec [here](https://microsoft.github.io/language-server-protocol/specifications/specification-current/) but still not quite get what's the purpose of it..

**Comment by Shatur95** [comment]
Oh, I'm sorry. I think that `commitCharacters` not related to my suggestion. This is just a characters that will trigger completion when pressed. [Here](https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.languageserver.protocol.completionitem.commitcharacters?view=visualstudiosdk-2019) is the explanation.

**Comment by haorenW1025** [comment]
Okay then I think I can make a custom option for that, something like `completion_expandCharacters` and you can define it as `" \t)>}"`, it should resolve this issue.

**Comment by Shatur95** [comment]
Yes, I think this will be the best way.

**Comment by Shatur95** [comment]
Closed by #151.
